### PR TITLE
api additions needed by scheduler

### PR DESCRIPTION
**Pull request by krancour**
Another PR that provides a fragment of scheduler functionality to avoid a much larger PR. This PR provides the API server endpoints and implementations for starting a worker or job (when requested to do so by the scheduler-- which comes in the next PR).

Note to reviewer: Don't be scurred. Most of the code is just test cases.

### CODE REVIEW

File: v2/apiserver/internal/core/kubernetes/substrate.go [hunk 1]
```diff
@@ -57,6 +59,17 @@ type SubstrateConfig struct {
 	// this information whenever it needs to tell another component where to find
 	// the API server.
 	APIAddress string
+	// DefaultWorkerImage is the name/version of the OCI image that will be used
```
Comment (radu-matei) [inline v2/apiserver/internal/core/kubernetes/substrate.go:62]: Could this comment be a bit more explicit about the expected formatting of the image name and version?

Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:62]: Sure. On it.

File: v2/apiserver/internal/core/kubernetes/substrate.go [hunk 2]
```diff
@@ -75,12 +105,17 @@ func NewSubstrate(
...
+	s := &substrate{
 		generateNewNamespaceFn: generateNewNamespace,
 		kubeClient:             kubeClient,
 		queueWriterFactory:     queueWriterFactory,
 		config:                 config,
 	}
+	s.createWorkspacePVCFn = s.createWorkspacePVC
+	s.createWorkerPodFn = s.createWorkerPod
+	s.createJobSecretFn = s.createJobSecret
+	s.createJobPodFn = s.createJobPod
```
Comment (radu-matei) [inline v2/apiserver/internal/core/kubernetes/substrate.go:117]: It took me a second to see the `Fn` at the end there.
(no changes required here, just pointing it out)

File: v2/apiserver/internal/core/kubernetes/substrate.go [hunk 3]
```diff
@@ -568,3 +655,622 @@ func (s *substrate) countRunningPods(
...
 }
+
+func (s *substrate) createWorkspacePVC(
+	ctx context.Context,
+	project core.Project,
+	event core.Event,
+) error {
+	storageQuantityStr := event.Worker.Spec.WorkspaceSize
+	if storageQuantityStr == "" {
+		storageQuantityStr = "1G"
```
Comment (radu-matei) [inline v2/apiserver/internal/core/kubernetes/substrate.go:666]: Should this be a constant somewhere?

Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:666]: Thought about it, but I don't know what benefit we'd get from it.

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 2 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:117]: Ya... will warn that I end up using this pattern in a lot of places. "Constructor" initializes everything correctly, but functions are individually overridable for test purposes. It ends up looking a little complicated, but it improves testability so much.

File: v2/apiserver/internal/core/kubernetes/substrate.go [hunk 4]
```diff
@@ -568,3 +655,622 @@ func (s *substrate) countRunningPods(
...
+
+		initContainers = append(
+			initContainers,
+			corev1.Container{
+				Name: "vcs",
+				// TODO: For now, we're using the Brigade 1.x git init image, but for
+				// the sake of consistency and to lower the bar for creating additional
+				// VCS integrations, we should develop a new/improved image that gets
+				// input from a chunk of JSON, just like the actual worker image does.
+				Image:           "brigadecore/git-sidecar:v1.4.0",
```
Comment (radu-matei) [inline v2/apiserver/internal/core/kubernetes/substrate.go:816]: Same comment here about a constant for the string.

Would there be any point in just tagging `1.4.0` as `2.0.0-alpha` or something?

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 1 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:62]: Would this work?

```
// DefaultWorkerImage is the name of the OCI image that will be used for the
// Worker pod's container[0] if none is specified in a Project's
// configuration. The expected format is [REGISTRY/][ORG/]IMAGE_NAME[:TAG].
```

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 4 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:816]: Nice catch. I should add a TODO comment in here, as this is temporary.

There's an open issue for this. See #1134

File: v2/apiserver/internal/core/kubernetes/substrate.go [hunk 5]
```diff
@@ -568,3 +655,622 @@ func (s *substrate) countRunningPods(
...
+		if sidecarSpec.UseSource {
+			containers[i].VolumeMounts = append(
+				containers[i].VolumeMounts,
+				corev1.VolumeMount{
+					Name:      "vcs",
+					MountPath: sidecarSpec.SourceMountPath,
+				},
+			)
+		}
+		if sidecarSpec.UseHostDockerSocket {
```
Comment (radu-matei) [inline v2/apiserver/internal/core/kubernetes/substrate.go:1222]: It feels like adding the specific volumes for sources, workspaces, and Docker sockets is done at least 3 times, IIRC.

Would it make sense to have some simple helper functions that do this?

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 4 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:816]: Wait... that comment already is a TODO.

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 5 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:1222]: I'll look at it. iirc, I tried that at one point and things were less readable, but it's very reasonable to give that another try.

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 1 above)
Comment (radu-matei) [reply v2/apiserver/internal/core/kubernetes/substrate.go]: Perfect!

**Review by radu-matei — APPROVED** [review]
Just a few non-blocking thoughts about code reuse, otherwise LGTM.

File: v2/apiserver/internal/core/kubernetes/substrate.go (hunk 5 above)
Comment (krancour) [reply v2/apiserver/internal/core/kubernetes/substrate.go:1222]: Yes. I can DRY this up nicely.

Thanks for encouraging me to do so.

**Pull request merged**